src/frontend/Promotion.ml

```diff
@@ -0,0 +1,98 @@
+open Core_kernel
+open Core_kernel.Poly
+module UnsizedType = Middle.UnsizedType
+
+(** Type to represent promotions in the typechecker.
+  This can be used to return information about how to promote
+  expressions for use in [Ast.Promotion]*)
+type t =
+  | NoPromotion
+  | IntToReal
+  | ToVar (* used in arrays, not functions *)
+  | ToComplexVar (* used in arrays, not functions *)
```

Comment:
I am not sure I understand what the comment `used in arrays, not functions` means? Can you give me an example?

Reply:
We don't (at the moment at least) have any functions which require Var arguments, so these will never be created inside `SignatureMismatch.ml`. But, when building this PR, you do run into situations like this

```stan
parameters {
complex z;
array[2] complex x = {z; 3};  
}
```

Where promoting the 3 to `complex<var>` is necessary (complex<double> would fail).
There are similar cases where we need to promote a double to a var - these don't exist in function calls, ever.

This is why the `promotion_cost` of these is 0, since that promotion cost is for function resolution

Reply:
Got it. Thanks!